packages/client/src/router.ts

```diff
@@ -34,11 +40,22 @@ export const createVueRouter = (): Router => {
 
   // ensure page data and page component have been loaded before resolving the route,
   // and save page data to route meta
-  router.beforeResolve(async (to, from) => {
+  router.beforeResolve(async (to, from): Promise<string | void> => {
     if (to.path !== from.path || from === START_LOCATION) {
+      const pagePath = resolvers.resolvePagePath(pagesMap, to.path)
+
+      if (pagePath !== to.path) {
+        return pagePath
+      }
+
+      // if no match at this point, then we should provide 404 page
+      const pageInfo = pagesMap.get(pagePath) || pagesMap.get('/404.html')!
+
+      // TODO: Added for backwards compatibility, remove in stable version
+      to.meta = pageInfo.meta
       ;[to.meta._data] = await Promise.all([
-        resolvers.resolvePageData(to.name as string),
-        pagesComponents[to.name as string]?.__asyncLoader(),
+        resolvers.resolvePageData(pageInfo, pagePath),
+        pageInfo.comp?.__asyncLoader(),
```

Comment:
Notes:

Since we use beforeResolve to align pageData with routePath, we should avoid avoid callindg router.resolve.

This means we should avoid using `router-link` as it internally calls `router.resolve` to provide correct link on anchor tag.

So https://github.com/vuepress/vuepress-next/pull/1353 is still needed.

Though we no longer need RouterLink in that PR, VPLink should still be used by default in markdown and navbar, sidebar links.

Reply:
@meteorlxy I prefer splitting the VPLink PR by  updating the existing one #1353 after you merge this.